docs/spec_v5_errors.md

```diff
@@ -0,0 +1,173 @@
+  not in a state required for the operation's execution. For example, doing
+  an rmdir operation on a directory that is non-empty, or on a non-directory
+  object, or when having conflicting read-modify-write on the same resource.
+* **aborted**: The operation was aborted, typically due to a concurrency
+  issue like sequencer check failures, transaction aborts, etc.
+* **out_of_range**: The operation was attempted past the valid range. 
+  For example, seeking or reading past end of a paginated collection.
+  Unlike "invalid_argument", this error indicates a problem that may be fixed if
+  the system state changes (i.e. adding more items to the collection).
+  There is a fair bit of overlap between "failed_precondition" and "out_of_range".
+  We recommend using "out_of_range" (the more specific error) when it applies so
+  that callers who are iterating through a space can easily look for an
+  "out_of_range" error to detect when they are done.
+* **unimplemented**: The operation is not implemented or not
+  supported/enabled in this service.
+* **internal**: When some invariants expected by the underlying system
+  have been broken. In other words, something bad happened in the library or
+  backend service. Twirp specific issues like wire and serialization problems
+  are also reported as "internal" errors.
+* **unavailable**: The service is currently unavailable. This is most
+  likely a transient condition and may be corrected by retrying with a
+  backoff.
+* **data_loss**: Unrecoverable data loss or corruption.
+
+
+### HTTP status code
+
+Twirp services must respond with status `200` for non-error responses.
+For error responses, the staus depends on the Twirp Error Code:
+
+```
+Twirp Error Code        HTTP status code
+---------------------- -----------------------
+"canceled"            | 408 RequestTimeout
+"unknown"             | 500 Internal Server Error
+"invalid_argument"    | 400 BadRequest
+"deadline_exceeded"   | 408 RequestTimeout
+"not_found"           | 404 Not Found
+"bad_route"           | 404 Not Found
+"already_exists"      | 409 Conflict
+"permission_denied"   | 403 Forbidden
+"unauthenticated"     | 401 Unauthorized
+"resource_exhausted"  | 403 Forbidden
+"failed_precondition" | 412 Precondition Failed
+"aborted"             | 409 Conflict
+"out_of_range"        | 400 Bad Request
+"unimplemented"       | 501 Not Implemented
+"internal"            | 500 Internal Server Error
+"unavailable"         | 503 Service Unavailable
+"dataloss"            | 500 Internal Server Error
+```
+
+### Non-Twirp Errors from Intermediary Proxies
+
+It is also possible for Twirp Clients to receive non-200 responses with invalid 
+Twirp Errors (non JSON, or with invalid keys), this responses may be
+generated by proxy middleware or network issues. Twirp Clients should build a
+valid Twirp error depending on the HTTP status of those invalid responses:
+
+```
+HTTP status code           Twirp Error Code
```

Comment:
I don't think we should specify this at the specification level. I don't see when it would need to be interoperable across languages - I think we can get away with saying that the choice here is an implementation detail.

It's like how HTTP doesn't specify that "status code" that you should report to clients if a TCP connection breaks. Similarly, we don't need to specify what clients should present if the remote end _isn't_ speaking Twirp. We should only cover what to do if it _is_ speaking Twirp.

However, this is still useful to document somewhere, maybe in `docs/errors.md`.

Reply:
Yeah I though about this, but I don't see why Twirp clients need to be inconsistent across languages and why should we let them re-discover the same issues over an over. Twirp doesn't only cover service and wire, but also auto-generated clients, so it made sense to include some notions about how to handle non-twirp responses.

In any case, it is true that this part makes the spec a little noisy, and I was using "should" vocabulary instead of "must". Maybe we should write a spec for Twirp Clients? Or a doc with recommendations? Or as you said, just including this in `docs/errors.md` could already work as a point of reference to other languages.

Reply:
Twirp's specification should be as small as possible. If this doesn't need to be in the spec, it shouldn't be.

Let's put this in a separate document like `docs/errors.md` to explain how it works in Go, but leave it up to language-specific implementers. It should be allowed to do this in a way that makes sense in their language.